spec/unit/request_registry_spec.rb

```diff
@@ -43,6 +43,25 @@
     it "should report number of times all matching pattern were requested" do
       expect(WebMock::RequestRegistry.instance.times_executed(WebMock::RequestPattern.new(:get, /.*example.*/))).to eq(3)
     end
+
+    describe "multithreading" do
+      let(:request_pattern) { WebMock::RequestPattern.new(:get, "www.example.com") }
+
+      # Reproduce a multithreading issue that causes a RuntimeError:
+      #   can't add a new key into hash during iteration.
+      it "works normally iterating on the requested signature hash while another thread is setting it" do
+        thread_injected = false
+        allow(request_pattern).to receive(:matches?).and_wrap_original do |m, *args|
+          unless thread_injected
+            thread_injected = true
+            Thread.new { WebMock::RequestRegistry.instance.requested_signatures.put(:abc) }.join(0.1)
```

Comment:
Had to `join` immediately to produce deterministic failure in the current implementation. However, it will also cause a deadlock in the proposed solution (since both `put` and `select` are waiting for the other to release the lock). Setting a limit by `join(0.1)` here addressed both cases.